src/base/static/components/story-sidebar/index.js

```diff
@@ -0,0 +1,190 @@
+          return memo.set(
+            chapterUrl,
+            fromJS(this.getModelFromUrl(chapterUrl).attributes),
+          );
+        }, OrderedMap()),
+      );
+    }, OrderedMap());
+
+    this.state = {
+      currentStoryHeader: "",
+      currentStoryDescription: "",
+      currentStoryName: "",
+      currentStoryChapters: OrderedMap(),
+      currentStoryRoute: "",
+      currentChapter: Map(),
+    };
+  }
+
+  componentWillMount() {
+    this.isInitialized = false;
+    this.loadAndSetStoryChapter(Backbone.history.getFragment().split("/"));
+    this.isInitialized = true;
+
+    this.props.router.on("route", (fn, route) => {
+      this.loadAndSetStoryChapter(route);
+    });
+  }
+
+  loadAndSetStoryChapter(route) {
+    let isRouteFound = false;
+    const joinedRoute = route.join("/");
+    this.stories.forEach((story, storyName) => {
+      if (story.get(joinedRoute)) {
+        // If the current route exists in this story, load the story and set
+        // the current chapter and name.
+        isRouteFound = true;
+        this.setState({
+          currentStoryHeader: storyConfig[storyName].header,
+          currentStoryDescription: storyConfig[storyName].description,
+          currentStoryName: storyName,
+          currentStoryChapters: story,
+          currentStoryRoute: joinedRoute,
+          currentChapter: story.get(joinedRoute),
+        });
+      }
+    });
+
+    // If we can't find a match for the current route and we haven't
+    // initialized the story sidebar, load content from the first story
+    // provided in the config so the sidebar is never without content.
+    if (!isRouteFound) {
+      this.setState({
+        currentStoryHeader: !this.isInitialized
+          ? storyConfig[this.stories.keySeq().first()].header
+          : "",
+        currentStoryDescription: !this.stories.isInitialized
+          ? storyConfig[this.stories.keySeq().first()].description
+          : "",
+        currentStoryName: !this.isInitialized
+          ? this.stories.keySeq().first()
+          : this.state.currentStoryName,
+        currentStoryChapters: !this.isInitialized
+          ? this.stories.first()
+          : this.state.currentStoryChapters,
+        currentStoryRoute: "",
+        currentChapter: Map(),
+      });
+    }
+  }
+
+  getModelFromUrl(url) {
+    const splitUrl = url.split("/");
+
+    // If the url has a slash in it with text on either side, we assume we have
+    // a place model url and can look up the model directly.
+    if (splitUrl.length === 2) {
+      return (
+        this.props.places[
+          this.props.layers.find(layer => layer.slug === splitUrl[0]).id
+        ].get(splitUrl[1]) || {}
+      );
+    } else {
+      // Otherwise, we have a "landmark-style" url, and have to search all place
+      // collections.
+      let foundModel;
+      Object.keys(this.props.places).forEach(collectionName => {
```

Comment:
I think we should use a `find` instead of a `forEach` here, ie:

`const foundModel = Object.keys(this.props.places).find(collectionName => {`

and have the callback return the `findWhere` result.

Reply:
I might be overlooking something, but I think using `find` and returning the result from `findWhere` will cause the result of the whole expression to be the collection name, not the model we're searching for.

I think we still need to "pluck" out the found model from within the callback to either `find` or `forEach`. Am I missing a more functional way to do this though?

Reply:
oh, I suppose we could `find` and then `map`. That seems more functional.